File: js/channels.js
// mauIRC - The original mauIRC web frontend
// Copyright (C) 2016 Tulir Asokan

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

"use strict"
function getActiveChannel() {
  "use strict"
  var active = $(".channel-switcher.active > .channel-switcher-name")
  if (active.length) {
    return active.text()
  }
  return ""
}

function getActiveNetwork() {
  "use strict"
  var active = $(".network-switcher.activenet > .network-switcher-name")
  if (active.length) {
    return active.text().trim()
  }
  return ""
}

function getActiveChannelObj() {
  "use strict"
  return $(".channel-container:not(.hidden)")
}

function getActiveNetworkObj() {
  "use strict"
  return $(".network-container:not(.hidden)")
}

function openPM(network, user) {
  "use strict"
  openChannel(network, user, true)
  switchTo(network, user)
}

function newChannel(network) {
  "use strict"
  network = network.toLowerCase()

  var oldAdder = $(sprintf("#channel-adder-%s", network))
  if (oldAdder.length !== 0) {
    oldAdder.focus()
    return
  }

  $(sprintf("#chanswitchers-%s", network)).loadTemplate($("#template-channel-adder"), {
    id: sprintf("channel-adder-%s", network),
    wrapid: sprintf("channel-adder-wrapper-%s", network),
    finish: sprintf("if (event.keyCode === 13) { finishNewChannel('%1$s') } else if (event.keyCode === 27) { cancelNewChannel('%1$s') }", network),
    blur: sprintf("cancelNewChannel('%s')", network)
  }, {append: true, isFile: false, async: false})
  $(sprintf("#add-channel-%s", network)).addClass("hidden")
  var adder = $(sprintf("#channel-adder-%s", network))
  adder.easyAutocomplete({
    data: data.getNetwork(network).getChannels(),
    placeholder: "Nick or #channel",
  	list: {
  		maxNumberOfElements: 10,
  		match: { enabled: true },
      sort: {enabled: true },
      onChooseEvent: function() { finishNewChannel(network) },
  		showAnimation: {
  			type: "slide",
  			time: 400,
  			callback: function() {}
  		},
  		hideAnimation: {
  			type: "slide",
  			time: 400,
  			callback: function() {}
  		}
  	}
  })
  adder.focus()
}

function cancelNewChannel(network) {
  "use strict"
  $(sprintf("#channel-adder-wrapper-%s", network)).remove()
  $(sprintf("#add-channel-%s", network)).removeClass("hidden")
}

function finishNewChannel(network) {
  "use strict"
  var adder = $(sprintf("#channel-adder-%s", network))
  if (adder.length === 0) {
    return
  }

  var name = adder.val().trim()
  $(sprintf("#channel-adder-wrapper-%s", network)).remove()

  if (name.length === 0) {
    return
  }

  if (name.startsWith("#")) {
    sendMessage({
      type: "message",
      network: network,
      channel: name,
      command: "join",
      message: "Joining"
    })
  }

  $(sprintf("#add-channel-%s", network)).removeClass("hidden")
  openChannel(network, name, true)
  switchTo(network, name)
}

function openNetwork(network) {
  "use strict"
  network = network.toLowerCase()
  if ($(sprintf("#net-%s", network)).length !== 0) {
    return
  }

  $("#messages").loadTemplate($("#template-network"), {
    network: sprintf("net-%s", network)
  }, {append: true, isFile: false, async: false})
  $("#networks").loadTemplate($("#template-network-switcher"), {
    network: sprintf("switchnet-%s", network),
    networkname: network,
    openchannel: sprintf("newChannel('%s')", network),
    networkbtns: sprintf("chanswitchers-%s", network),
    addchanid: sprintf("add-channel-%s", network)
  }, {append: true, isFile: false, async: false})
  createRawIO(network)
}

function openChannel(network, channel, byUser) {
  "use strict"
  network = network.toLowerCase()
  var chanLower = channel.toLowerCase()
  var netObj = $(sprintf("#net-%s", network))
  if (netObj.length === 0) {
    openNetwork(network)
    netObj = $(sprintf("#net-%s", network))
  }

  if ($(sprintf("#chan-%s-%s", network, channelFilter(channel))).length !== 0) {
    return
  }

  if (byUser) {
    sendMessage({
      type: "open",
      network: network,
      channel: channel
    })
  }

  netObj.loadTemplate($("#template-channel"), {
    channel: sprintf("chan-%s-%s", network, chanLower)
  }, {append: true, isFile: false, async: false})
  $(sprintf("#chanswitchers-%s", network)).loadTemplate($("#template-channel-switcher"), {
    channel: sprintf("switchto-%s-%s", network, chanLower),
    brid: sprintf("break-chan-%s-%s", network, chanLower),
    channelname: channel,
    onclick: sprintf("switchTo('%s', '%s')", network, chanLower)
  }, {append: true, isFile: false, async: false})
}

function closeChannel(network, channel) {
  "use strict"
  network = network.toLowerCase()
  var netObj = $(sprintf("#net-%s", network))
  if (netObj.length === 0) {
    return
  }

  var chanObj = $(sprintf("#chan-%s-%s", network, channelFilter(channel)))
  if (chanObj.length === 0) {
    return
  }

  if (getActiveNetwork() === network && getActiveChannel() == channel) {
    switchToClear()
  }
  chanObj.remove()
  $(sprintf("#switchto-%s-%s", network, channelFilter(channel))).remove()
  $(sprintf("#break-chan-%s-%s", network, channelFilter(channel))).remove()
}

function switchView(userlist) {
  "use strict"
  if ($("#messaging").hasClass("hidden-tablet-down")) {
    if ($("#userlist").hasClass("hidden-tablet-down")) {
      $("#networks").addClass("hidden-tablet-down")
    } else {
      $("#userlist").addClass("hidden-tablet-down")
    }
    $("#messaging").removeClass("hidden-tablet-down")
  } else {
    if (userlist) {
      $("#userlist").removeClass("hidden-tablet-down")
      $("#messaging").addClass("hidden-tablet-down")
    } else {
      $("#networks").removeClass("hidden-tablet-down")
      $("#messaging").addClass("hidden-tablet-down")
    }
  }
}

function switchToClear() {
  "use strict"
  getActiveChannelObj().addClass("hidden")
  getActiveNetworkObj().addClass("hidden")
  $(".channel-switcher.active").removeClass("active")
  $(".network-switcher.activenet").removeClass("activenet")
  $("#title").text("")
  if ($("#messaging").hasClass("hidden-tablet-down")) {
    switchView(false)
  }
}

function switchTo(network, channel) {
  "use strict"
  network = network.toLowerCase()
  var chanFiltered = channelFilter(channel)
  dbg(sprintf("Switching to channel %s @ %s", channel, network))
  getActiveChannelObj().addClass("hidden")
  getActiveNetworkObj().addClass("hidden")
  $(".channel-switcher.active").removeClass("active")
  $(".network-switcher.activenet").removeClass("activenet")
  $("#message-text").focus()

  var channelData = data.getChannel(network, channel)
  if (channelData.getTopic().length !== 0) {
    var title = channelData.getTopic()
  } else {
    var title = channel
  }
  $("#title").text(title)

  if ($("#messaging").hasClass("hidden-tablet-down")) {
    switchView(false)
  }

  if (!channelData.isHistoryFetched()) {
    history(network, channel, 512)
    channelData.setHistoryFetched()
  }

  $(sprintf("#switchnet-%s", network)).addClass("activenet")
  $(sprintf("#net-%s", network)).removeClass("hidden")
  $(sprintf("#chan-%s-%s", network, chanFiltered)).removeClass("hidden")

  var newChanSwitcher = $(sprintf("#switchto-%s-%s", network, chanFiltered))
  newChanSwitcher.removeClass("new-messages")
  newChanSwitcher.addClass("active")

  updateUserList()
  scrollDown()
}

File: ui/channels.go
// mauIRC - The original mauIRC web frontend
// Copyright (C) 2016 Tulir Asokan

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// Package ui contains UI-related functions
package ui

import (
	"fmt"
	"github.com/gopherjs/jquery"
	"maunium.net/go/mauirc/data"
	"maunium.net/go/mauirc/templates"
	"regexp"
	"strings"
)

// GetActiveChannel gets the name of the active channel
func GetActiveChannel() string {
	active := jq(".channel-switcher.active > .channel-switcher-name")
	if active.Length > 0 {
		return active.Text()
	}
	return ""
}

// GetActiveNetwork gets the name of the active network
func GetActiveNetwork() string {
	active := jq(".network-switcher.activenet > .network-switcher-name")
	if active.Length > 0 {
		return strings.TrimSpace(active.Text())
	}
	return ""
}

// GetActiveChannelObj gets the active channel container
func GetActiveChannelObj() jquery.JQuery {
	return jq(".channel-container:not(.hidden)")
}

// GetActiveNetworkObj gets the active network container
func GetActiveNetworkObj() jquery.JQuery {
	return jq(".network-container:not(.hidden)")
}

// GetChannel returns the container for the channel with the given name
func GetChannel(network, channel string) jquery.JQuery {
	return jq(fmt.Sprintf("#chan-%s-%s", NetworkFilter(network), ChannelFilter(channel)))
}

// GetNetwork returns the container for the network with the given name
func GetNetwork(network string) jquery.JQuery {
	return jq(fmt.Sprintf("#net-%s", NetworkFilter(network)))
}

var chanfilter = regexp.MustCompile("([\\#\\*\\.])")

// ChannelFilter filters the given channel name into a string usable in jQuery
func ChannelFilter(channel string) string {
	return strings.ToLower(chanfilter.ReplaceAllString(channel, "\\$1"))
}

// NetworkFilter filters the given network name into a string usable in jQuery
func NetworkFilter(network string) string {
	return strings.ToLower(network)
}

// OpenChannelAdder opens the channel adder for the given network
func OpenChannelAdder(network string) {
	network = NetworkFilter(network)

	var oldAdder = jq(fmt.Sprintf("#channel-adder-%s", network))
	if oldAdder.Length != 0 {
		oldAdder.Focus()
		return
	}

	templates.Append("channel-adder", fmt.Sprintf("#chanswitchers-%s", network), map[string]interface{}{
		"ID":     fmt.Sprintf("channel-adder-%s", network),
		"WrapID": fmt.Sprintf("channel-adder-wrapper-%s", network),
		"Finish": fmt.Sprintf("if (event.keyCode === 13) { ui.finishChannelAdding('%[1]s') } else if (event.keyCode === 27) { ui.cancelChannelAdding('%[1]s') }", network),
		"Blur":   fmt.Sprintf("ui.cancelChannelAdding('%s')", network),
	})

	adder := jq(fmt.Sprintf("#channel-adder-%s", network))
	adder.Focus()
	/* TODO: Autocomplete? Original JS implementation:
	   adder.easyAutocomplete({
	     data: data.getNetwork(network).getChannels(),
	     placeholder: "Nick or #channel",
	     list: {
	       maxNumberOfElements: 10,
	       match: { enabled: true },
	       sort: {enabled: true },
	       onChooseEvent: function() { finishNewChannel(network) },
	       showAnimation: {
	         type: "slide",
	         time: 400,
	         callback: function() {}
	       },
	       hideAnimation: {
	         type: "slide",
	         time: 400,
	         callback: function() {}
	       }
	     }
	   })
	*/
}

// CancelChannelAdding closes the channel adder
func CancelChannelAdding(network string) {
	jq(fmt.Sprintf("#channel-adder-wrapper-%s", network)).Remove()
	jq(fmt.Sprintf("#add-channel-%s", network)).RemoveClass("hidden")
}

// FinishChannelAdding finishes adding a channel
func FinishChannelAdding(network string) {
	adder := jq(fmt.Sprintf("#channel-adder-%s", network))
	if adder.Length == 0 {
		return
	}

	name := strings.TrimSpace(adder.Val())
	jq(fmt.Sprintf("#channel-adder-wrapper-%s", network)).Remove()
	if len(name) == 0 {
		return
	}

	if name[0] == '#' {
		/* TODO how to send messages?
		  data.Messages <- messages.Container{
			  Type: messages.MsgMessage,
			  Object: messages.Message{
				  Network: network,
				  Channel: name,
				  Command: "join",
				  Message: "Joining",
			  },
		  }
		*/
	}

	jq(fmt.Sprintf("#add-channel-%s", network)).RemoveClass("hidden")
	OpenChannel(network, name, true)
	SwitchTo(network, name)
}

// OpenNetwork opens a network
func OpenNetwork(network string) {
	network = NetworkFilter(network)

	if jq(fmt.Sprintf("#net-%s", network)).Length != 0 {
		return
	}

	templates.Append("network", "#messages", network)
	templates.Append("network-switcher", "#networks", network)
	// CreateRawIO(network)
}

// OpenChannel opens a channel
func OpenChannel(network, channel string, byUser bool) {
	network = NetworkFilter(network)
	chanLower := strings.ToLower(channel)

	netObj := GetNetwork(network)
	if netObj.Length == 0 {
		OpenNetwork(network)
		netObj = GetNetwork(network)
	}

	if GetChannel(network, channel).Length != 0 {
		return
	}

	if byUser {
		/* TODO send messages?
		data.Messages <- messages.Container{
			Type: messages.MsgOpen,
			Object: messages.Open{
				Network: network,
				Channel: channel,
			},
		}
		*/
	}

	templates.AppendObj("channel", netObj, fmt.Sprintf("chan-%s-%s", network, chanLower))
	templates.Append("channel-switcher", fmt.Sprintf("#chanswitchers-%s", network), map[string]interface{}{
		"Channel":     chanLower,
		"ChannelReal": channel,
		"Network":     network,
	})
}

// CloseChannel closes a channel
func CloseChannel(network, channel string) {
	network = NetworkFilter(network)
	chanFiltered := ChannelFilter(channel)

	netObj := GetNetwork(network)
	if netObj.Length == 0 {
		return
	}

	chanObj := GetChannel(network, channel)
	if chanObj.Length == 0 {
		return
	}

	if GetActiveNetwork() == network && GetActiveChannel() == channel {
		SwitchToClear()
	}
	chanObj.Remove()
	jq(fmt.Sprintf("#switchto-%s-%s", network, chanFiltered)).Remove()
	jq(fmt.Sprintf("#break-chan-%s-%s", network, chanFiltered)).Remove()
}

// SwitchToClear switches to a empty message view
func SwitchToClear() {
	GetActiveChannelObj().AddClass("hidden")
	GetActiveNetworkObj().AddClass("hidden")
	jq(".channel-switcher.active").RemoveClass("active")
	jq(".network-switcher.activenet").RemoveClass("activenet")
	jq("#title").SetText("")
	OpenMessageView()
}

// SwitchTo switches to the given channel on the given network
func SwitchTo(network, channel string) {
	network = NetworkFilter(network)
	fmt.Printf("Switching to channel %s @ %s\n", channel, network)

	SwitchToClear()
	jq("#message-text").Focus()

	var title string
	chanData := data.Networks.MustGetChannel(network, channel)
	if len(chanData.Topic) != 0 {
		title = chanData.Topic
	} else {
		title = channel
	}
	jq("#title").SetText(title)

	chanObj := GetChannel(network, channel)
	if chanData.HistoryFetched && chanObj.Find(".invite-wrapper").Length == 0 {
		// TODO fetch history
		// history(network, channel, 512)
	}

	jq("#switchnet-%s", network).AddClass("activenet")
	jq("#net-%s", network).RemoveClass("hidden")
	chanObj.RemoveClass("hidden")
	chanSwitcher := jq(fmt.Sprintf("#switchto-%s-%s", network, ChannelFilter(channel)))
	chanSwitcher.RemoveClass("new-messages")
	chanSwitcher.AddClass("active")
	UpdateUserlist()
	ScrollDown()
}

// OpenPM opens a private query with the given user on the given network
func OpenPM(network, user string) {
	OpenChannel(network, user, true)
	SwitchTo(network, user)
}

// OpenMessageView switches to the message view
func OpenMessageView() {
	jq("#networks").AddClass("hidden-medium-down")
	jq("#messaging").RemoveClass("hidden-medium-down")
	jq("#userlist").AddClass("hidden-medium-down")
}

// OpenNetworksView switches to the network list view
func OpenNetworksView() {
	jq("#networks").RemoveClass("hidden-medium-down")
	jq("#messaging").AddClass("hidden-medium-down")
	jq("#userlist").AddClass("hidden-medium-down")
}

// OpenUserlistView switches to the userlist view
func OpenUserlistView() {
	jq("#networks").AddClass("hidden-medium-down")
	jq("#messaging").AddClass("hidden-medium-down")
	jq("#userlist").RemoveClass("hidden-medium-down")
}

// ToggleNetworksView toggles between the network and message view
func ToggleNetworksView() {
	if jq("#networks").HasClass("hidden-medium-down") {
		OpenNetworksView()
	} else {
		OpenMessageView()
	}
}

// ToggleUserlistView toggles between the userlist and message view
func ToggleUserlistView() {
	if jq("#userlist").HasClass("hidden-medium-down") {
		OpenUserlistView()
	} else {
		OpenMessageView()
	}
}
